apps/src/lib/proto/types.rs

```diff
@@ -0,0 +1,329 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+use prost::Message;
+use thiserror::Error;
+
+use super::generated::{services, types};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error decoding a transaction from bytes: {0}")]
+    TxDecodingError(prost::DecodeError),
+    #[error("Error decoding an IntentGossipMessage from bytes: {0}")]
+    IntentDecodingError(prost::DecodeError),
+    #[error("Error decoding an DkgGossipMessage from bytes: {0}")]
+    DkgDecodingError(prost::DecodeError),
+    #[error("Intent is empty")]
+    NoIntentError,
+    #[error("Dkg is empty")]
+    NoDkgError,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub struct Tx {
+    inner: types::Tx,
+}
+
+impl Tx {
+    pub fn from(tx_bytes: impl AsRef<[u8]>) -> Result<Self> {
+        let inner = types::Tx::decode(tx_bytes.as_ref())
+            .map_err(Error::TxDecodingError)?;
+        Ok(Tx { inner })
+    }
+
+    pub fn new(code: Vec<u8>, data: Option<Vec<u8>>) -> Self {
+        let inner = types::Tx {
+            code,
+            data,
+            timestamp: Some(std::time::SystemTime::now().into()),
+        };
+        Tx { inner }
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = vec![];
+        self.inner
+            .encode(&mut bytes)
+            .expect("encoding a transaction failed");
+        bytes
+    }
+
+    pub fn code(&self) -> Vec<u8> {
+        self.inner.code.clone()
+    }
+
+    pub fn data(&self) -> Vec<u8> {
+        match &self.inner.data {
+            Some(d) => d.clone(),
+            None => Vec::new(),
+        }
+    }
```

Comment:
I think it's a bit strange to returns an empty vec when there is no data, it might be better instead to return `Option<Vec<u8>>`

When using this function it should be checkable if the tx has data or not

Reply:
That makes sense! Thank you!